shared-libs/task-utils/src/task-utils.js

```diff
@@ -0,0 +1,33 @@
+var matchedLastHistory = function(task, state) {
+  if (!task.state_history.length) {
+    return false;
+  }
+
+  return task.state_history[task.state_history.length - 1].state === state;
+};
+
+var setTaskState = function(task, state, details) {
+  var timestamp = (new Date()).toISOString();
+
+  task.timestamp = timestamp;
+  task.state_history = task.state_history || [];
+
+  if (task.state === state && (!details || task.state_details === details) && matchedLastHistory(task, state)) {
+    task.state_history[task.state_history.length - 1].timestamp = timestamp;
+    return false;
+  }
+
+  task.state = state;
+  task.state_details = details;
+  task.state_history.push({
+    state: state,
+    state_details: details,
+    timestamp: timestamp
+  });
+
+  return true;
+};
+
+module.exports = {
+  setTaskState: setTaskState
```

Comment:
Add some JS doc documentation here because this is now a library rather than just an inline function. Something explaining what the boolean return actually means would be really useful.

I _think_ the boolean return means whether or not it was updated but on line 16 we update the history timestamp and then return false which seems really strange to me. I know you're probably just sticking with the existing implementation but now's a good time to clean it up and work out how/if it works.

Reply:
I've updated the code to not update timestamps if not changing the state - so now the return value should be clear. 
I thought it was more useful to know when the last attempted state change happened, rather than the first.